libs/sfm/ba_sparse_matrix.h

```diff
@@ -299,6 +314,88 @@ SparseMatrix<T>::multiply (SparseMatrix const& rhs) const
     return ret;
 }
 
+template <typename T>
+SparseMatrix<T>
+SparseMatrix<T>::parallel_multiply (SparseMatrix const& rhs) const
+{
+    if (this->cols != rhs.rows)
+        throw std::invalid_argument("Incompatible matrix dimensions");
+
+    std::size_t nnz = this->num_non_zero() + rhs.num_non_zero();
+    SparseMatrix ret(this->rows, rhs.cols);
+    ret.reserve(nnz);
+    std::fill(ret.outer.begin(), ret.outer.end(), 0);
+
+    std::size_t const chunk_size = 64;
+    std::size_t const num_chunks = ret.cols / chunk_size
+         + (ret.cols % chunk_size != 0);
+    std::size_t const max_threads = std::max(1u,
+        std::thread::hardware_concurrency());
```

Comment:
It appears that using all available threads does not make sense on CPUs with hyper threading but I could not find a way to determine the number of physical cores.